### ISSUE TITLE
[Page|Frame].WaitForSelector is flaky

### ISSUE THREAD

**Issue by inancgumus** [grafana/xk6-browser#258]
_This issue is valid for [v0.2.0](https://github.com/grafana/xk6-browser/releases/tag/v0.2.0)._

While working on #171, I noticed that `waitForSelector` is fragile and may randomly fail if the script doesn't sleep or slow-motion between steps.

<details>
<summary>The flaky script.</summary>

```js
import { group } from 'k6';
import launcher from 'k6/x/browser';

export default function () {
  const browser = launcher.launch('chromium', {
    headless: false,
    // slowMo: '500ms' // slow down by 500ms
  });

  let context = browser.newContext({
    // hack:
    // Without setting viewport, the script cannot find the add-to-cart
    // element. We might want to improve this.

    // screen: { width: 1920, height: 1080 },
    // viewport: { width: 1920, height: 1080 },
  });

  const page = context.newPage();

  group("Navigate to Homepage", () => {
    page.goto('http://ecommerce.test.k6.io/', { waitUntil: 'networkidle' });

    page.waitForSelector('.woocommerce-result-count');
  });

  group("Click Product", () => {
    page.click('.woocommerce-loop-product__title', { waitUntil: 'networkidle' });
  });

  group("Click Add to Cart", () => {
    console.log("Clicking Add to Cart...");

    // wait for the element to be attached
    let b = page.waitForSelector('button[name="add-to-cart"]');

    // hack: scroll the element into view
    page.evaluate(() => {
      const el = document.querySelector('button[name="add-to-cart"]');
      el.scrollIntoView();
    });
    b.click();

    console.log("Waiting for .woocommerce-message...");

    page.waitForSelector('.woocommerce-message');
  });

  group("View Cart", () => {
    page.click('.cart-contents', { waitUntil: 'networkidle' });

    page.waitForSelector('.woocommerce-cart-form__cart-item');
  });

  page.close();
  browser.close();
}
```

</details>

However, this may be the tip of the iceberg because the problem is more profound: Stale execution contexts #259.

**Explanation:**

* Frame calls `Runtime.callFunctionOn(...)` using execution context: X.
* Frame navigates.
* New execution context gets created: Y.
* FrameSession updates the new execution context Y for the frame.
* `Runtime.callFunctionOn` returns an error because it evaluated the function call in the execution context X instead of Y.

One solution is retrying after a failure, say, five times or so. Doing so fixed the problem (I ran the same script several times). How to test this is whole another problem. Let me know if you have any ideas? @imiric

**Comment by imiric** [comment]
Could this be part of #200?

**Comment by inancgumus** [comment]
Is not #200 about racing `waitForXXX`s? I mean they don't wait for the next action to happen. But, 🤷 , could you give it a try and see whether it fixes #200 as well?

**Comment by imiric** [comment]
It sounds like the same issue to me. They also needed to add `sleep()` for it to work.

Though it's difficult to test their specific scenario since this issue is entirely dependent on the site. For example, I didn't run into it at all while writing the script for our internal staging site.

But OK, let's leave it separate. #200 probably should've been split per method anyway.

**Comment by imiric** [comment]
Hey, I haven't been able to reproduce this on `main`, but I think I see how it can happen.

Correct me if I'm wrong, but this is a race condition between any event that causes navigation (clicking, form submission, etc.) and any `waitFor*` method?

That is, we need to wait for navigation to complete before doing any page interactions within that `ExecutionContext`. If navigation happens while waiting for a specific element, then we have no guarantees the interactions will be valid within the new `ExecutionContext`. Retrying makes some sense to handle the race condition, but it's a brute force method that could cause unnecessary CDP traffic if the element doesn't exist anyway when the navigation completes. Given that it's easy to mistype selectors, it would also introduce a delay of 5x whatever the configured timeout was, only to fail anyway.

So I agree with your suggestion in [this comment](https://github.com/grafana/xk6-browser/pull/260#discussion_r816062142). The better approach would be to cancel the actions, and try them with the new `ExecutionContext` once, and if _that_ fails, fail altogether. And yeah, it doesn't sound trivial :sweat_smile:

Re: testing, it should be relatively straightforward to reproduce an async navigation while `waitForSelector()` is running, right? Not in JS code before we have event loops, but in our Go integration tests.

**Comment by inancgumus** [comment]
This issue happens 3/5 of all test runs on my machine 🤔 I also have a test that can reproduce this issue, but there is a race condition (#261) around there, so I reverted the test from #260. Yes, it's a race condition but IDK how far it goes. So far, I can confirm that it can happen with anything that depends on execution contexts (e.g., ElementHandle).

> Although we don't know how many navigations and ExecutionContexts will be needed until the element is expected to exist.

I found that retrying only once is enough to make an _execution context refresh_. So we don't need to make it five times. Actions running on an older execution context keep running if we have a new execution context—retrying only once solves the problem. This issue wouldn't have occurred if the `Click` didn't happen before the navigation ends:

```js
// waitForSelect('.cart-contents') below starts running!
page.waitForNavigation({waitUntil: 'load'});
// a new execution context gets created 
const el = page.waitForSelector('.cart-contents');
// ^-- this line should have executed before the navigation
```

> So maybe we should expose the retry argument to JS anyway, except it shouldn't take into account timeout but consecutive context changes, i.e. navigations? 🤔

This sounds like #100. Check it out.

> ...If a navigation happens I'm now more in favor of just cancelling any existing waits and returning an error...

Agree with all your points. However, we can follow the opportunistic path here: I'm on the side of providing this band-aid solution until we settle on a permanent solution. A permanent solution is to wait until navigation so we need #200 for that. 

> Re: testing, it should be relatively straightforward to reproduce an async navigation while waitForSelector() is running, right? Not in JS code before we have event loops, but in our Go integration tests.

Yes, I tried something similar but the test isn't realistic because a single VU should run both navigation and click actions. If we use another goroutine to do that, it's like having multiple VUs operating on the same script. But I'll give it another try.

**Comment by imiric** [comment]
> This issue wouldn't have occurred if the `Click` didn't happen before the navigation ends

Aha, so this is caused by `waitForNavigation()` "not working", i.e. #200.

> This sounds like #100.

Yeah, kind of. I was just brainstorming on how it could work for this particular fix, and probably should've deleted that. :) I'm not sure we should expose `retry`, but instead do the workaround without retries, or fix the underlying issue.


> I'm on the side of providing this band-aid solution until we settle on a permanent solution.

I'm on the fence about this particular workaround, since it has drawbacks for cases when a wrong selector is used, like I mentioned above. If you can work on detecting when the `ExecutionContext` changes, that would be better, but I feel like we're stepping on each other's toes here since I'm also looking into #200 again.


> Yes, I tried something similar but the test isn't realistic because a single VU should run both navigation and click actions. If we use another goroutine to do that, it's like having multiple VUs operating on the same script.

Well, browser events happen asynchronously, and we do handle them as such in Go, so it would be replicating what essentially happens.

**Comment by inancgumus** [comment]
> ...instead, do the workaround without retries, or fix the underlying issue.
> If you can work on detecting when the ExecutionContext changes, that would be better...

Yeah, you'll probably fix this in #200 😄 But in the meantime, I see no reason for fixing it with a single retry 🤷  Even though there is a wrong selector, this would only take one more timeout cycle, and I think it's a fair cost to pay for making the selectors less fragile. I'd like to merge this fix and continue working on #171 😅  You can provide a better fix if you like in #200 ❤️ 

> Well, browser events happen asynchronously, and we do handle them as such in Go, so it would be replicating what essentially happens.

Yeah, they happen asynchronously, but `waitForSelector` and `Click` should synchronously, no? Isn't that what #200 is all about?

**Comment by inancgumus** [comment]
Re: Wrong selectors. We can retry for the _"Cannot find context with specified id"_ error, so those wrong selectors won't pass.
